### ISSUE TITLE
Enforcing data constraints after merge

### ISSUE THREAD

**Issue by saranrapjs** [automerge/automerge#158]
Hi! I've been giving an extremely rough go at [wiring up automerge to ProseMirror](https://discuss.prosemirror.net/t/prosemirror-crdts/1190/9), a rich-text editing library. (I also saw you speak at my work, the NYT, many moons ago — very cool to see this library develop!)

An issue that's come up is that ProseMirror has a way of expressing constraints on the data structures it allows for a given document using something called a [schema](http://prosemirror.net/docs/guide/#schema). This is oversimplifying it, but it allows you to express rules like "only allow a maximum of three paragraphs" or "allow two list elements when preceded by a blockquote" using a regex-like mini DFA language.

I'm not quite there with my experiment yet, but at some point I'll need to reconcile the fact that automerge makes a best guess as to how to converge on a state given adjacent changes originating from multiple automerge docs, and that these guesses will occasionally lead to invalid states for a given ProseMirror schema. It may bear noting here that ProseMirror document structures are ordered trees, e.g. something like:

```
Node:
  content: [?Node]
  attrs: Object,
```

...and the way I'm translating this to an automerge document is as a list/array (for the purposes of this issue, I'm only interested in changes to lists/arrays).

What I'm wondering: have you ever considered exposing some API that clients could share, which would allow end-users to specify a custom merge strategy? e.g. where, in cases like mine, you could enforce a rule like "this list may only contain a maximum of X children"? I read [this issue about matrix CRDT's](https://github.com/automerge/automerge/issues/102), which feels related but maybe not identical. I'd appreciate any thoughts/ideas anyone has :)

**Comment by canadaduane** [comment]
I wonder if you could use the DSL to generate a mapping from constraints to custom data types.

Taking your example of "only allow a maximum of three paragraphs," you could express this (and the absence of the constraint) as a set of two possible data types:

1. 1-MAX(N) Paragraphs
2. Unlimited Paragraphs

In the case of (1) you could map to an Automerge.Table type, where paragraphs 1 thru N are columns. In the case of (2) you could map to a regular list.

Basically, you would be creating the constraints in the data type itself, rather than applying a constraint afterward. In functional programming circles I've heard this expressed as "make impossible states impossible."

**Comment by ept** [comment]
Hi @saranrapjs, sorry for the slow reply. I am happy to consider and explore data constraints and/or custom merge strategies, but currently don't yet have a clear idea of what this might look like. Would the approach outlined by @canadaduane work for you? (i.e. representing a list of up to 3 paragraphs not by an actual list object, but by a map with keys "1", "2" and "3".)

The trouble with custom merge functions is that they offer lots of room for shooting yourself in the foot, e.g. if the function is not commutative and associative as required, you might end up with different users in inconsistent states. A solution here might be to not allow a full Turing-complete language for merging, but supporting only restricted, declaratively defined merging rules, for which we can guarantee convergence. But we would need to hear about more potential use cases in order to figure out what this kind of rule language might look like.

Another idea I have discussed with @pvh is that some kinds of changes to a document might be better represented as an [event sourcing-style event](https://docs.microsoft.com/en-us/azure/architecture/patterns/event-sourcing) rather than a CRDT state mutation, and to define a [Redux-style reducer](https://redux.js.org/basics/reducers) function that turns events into state updates. Such a reducer could potentially enforce lots more application-specific consistency properties that a CRDT cannot express. However, it is so far unclear what the interplay between CRDTs and such an event sourcing approach should look like.

**Comment by ept** [comment]
Closing this issue for now as we don't currently have a concrete proposal for what custom merge functions should look like. Feel free to reopen if new ideas emerge.
